[inductor] Change AOTInductor to return output tensors

desertfire opened the pull request:
Summary:
Change AOTInductor to directly return output tensors instead of taking pre-allocated output tensors to return the results. This gives several benefits:

* It makes sure AOTInductor has the same behavior when managing the output tensors as the default Inductor, which is widely tested and thus more reliable.
* As we have debugged before, there are cases we still have to codegen extra copy_ ops to fill the pre-allocated output tensors which doesn't make sense for performance.
* With the coming enhanced memory planning, this again will make sure the memory planning logic is the between AOTInductor and Inductor, which will greatly simplify the problem and improve the reliability.

This change also combines D49494954 from Yang and https://github.com/pytorch/pytorch/pull/109560 from Angela.

Differential Revision: D49502318




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

desertfire commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -364,6 +341,8 @@ class AOTInductorModelBase {
 
   std::vector<ParamInfo> inputs_info_;
   std::vector<ParamInfo> outputs_info_;
+  std::vector<AtenTensorHandle> output_handles_;
```
Didn't remove this. Will do in the next PR update.

chenyang78 commented on test/cpp/aot_inductor/test.cpp:
```diff
@@ -55,46 +55,36 @@ TEST(AotInductorTest, BasicTest) {
-  std::vector<torch::Tensor> outputs;
-  outputs.push_back(output_tensor);
+  auto input_handles =
+      torch::aot_inductor::unsafe_alloc_new_handles_from_tensors(input_tensors);
```
Hmm, do we need to change the var name at line 48 to ```input_tensors```?

desertfire replied:
Good catch. I should have run this cpp test locally.

chenyang78 commented on test/cpp/aot_inductor/test.cpp:
```diff
@@ -55,46 +55,36 @@ TEST(AotInductorTest, BasicTest) {
-  ASSERT_EQ(output_sizes[0][1], 10);
+      proxy_executor_handle));
+
+  outputs = torch::aot_inductor::alloc_tensors_by_stealing_from_handles(
```
Unless I am missing the place, seems we don't declare outputs yet? If that's the case, perhaps change it to ```auto outputs```?

desertfire replied:
Running this cpp test locally.

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1207,7 +1195,7 @@ def write_wrapper_decl(self):
                     c10::optional<at::Layout> optional_layout;
                     c10::optional<at::Tensor> optional_tensor;
-                    torch::List<c10::optional<at::Scalar>> optional_list;
+                    c10::List<c10::optional<at::Scalar>> optional_list;
```
dumb question - why are we switching to ```c10::List```?

desertfire replied:
It was coming my other include file change. Just for uniformity.

chenyang78 reviewed (approved):
Left some minor comments. LGTM. Thanks!

desertfire committed:
[inductor] Change AOTInductor to return output tensors (#109790)

Summary:

Change AOTInductor to directly return output tensors instead of taking pre-allocated output tensors to return the results. This gives several benefits:

* It makes sure AOTInductor has the same behavior when managing the output tensors as the default Inductor, which is widely tested and thus more reliable.
* As we have debugged before, there are cases we still have to codegen extra copy_ ops to fill the pre-allocated output tensors which doesn't make sense for performance.
* With the coming enhanced memory planning, this again will make sure the memory planning logic is the between AOTInductor and Inductor, which will greatly simplify the problem and improve the reliability.

This change also combined D49494954 from Yang.

Reviewed By: chenyang78

Differential Revision: D49502318
